man/man3/readconfig.3
.TH readconfig 3  "LPJmL programmers manual"
.SH NAME
readconfig \- Reads LPJmL configuration
.SH SYNOPSIS
.nf
\fB#include "lpj.h"

Bool readconfig(Config *\fIconfig\fB,
                Fscanpftparfcn \fIscanfcns\fB[],int \fIntypes\fB,
                int \fInout\fB,int *\fIargc\fB,char ***\fIargv\fB,
                const char *\fIusage\fB);\fP

.fi
.SH DESCRIPTION
Function reads LPJmL configuration.
The arguments of \fBreadconfig\fP are:
.TP
.I config
Pointer to LPJmL configuration data. Must be initialized by \fBinitmpiconfig()\fP or \fBinitconfig()\fP.
.TP 
.I scanfcns
Vector of scan functions for PFT-specific parameters.
.TP
.I ntypes
Number of plant functional type (PFT) classes.
.TP
.I nout
Maximum number of output files.
.TP
.I argc
Pointer to the number of arguments.
.TP
.I argv
Pointer to argument vector.
.TP
.I usage
Usage string printed in case of error in argument list or NULL.

.SH RETURN VALUE
Upon successful completion FALSE is returned and TRUE in case of error.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
initmpiconfig(3), initconfig(3), freeconfig(3)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.7.2] - 2023-08-11

### Added 
- outputs:
  - PFT_WATER_DEMAND: annual PFT specific water demand
  - AUTOTROPHIC_RESPIRATION: total annual autotrophic respiration
  - NDEPOS: total annual N deposition as sum of NH4 and NO3 deposition

## [5.7.1] - 2023-08-10

### Fixed

- wrong use of `&f` instead of `f`, in `fscanparam.c`. Bug was introduced in version 5.6.26 and caused that parameter files could not be read properly

## [5.7.0] - 2023-08-10

### Added

- land fraction can now be read from file, with a new flag `"landfrac_from_file"` and a new input `"landfrac"`
- If setting `"landfrac_from_file : true"` cell areas in LPJmL are scaled down by land fraction from input file. If land fraction is zero, land fraction is set to `minlandfrac` and a warning is printed.
- new flag `"with_lakes"` defined
- new outputs `"terr_area"`, `"land_area"` and `"lake_area"` defined writing the terrestrial area (land and lakes),  land area and lake area of each cell
- new parameters `"minlandfrac"` and `"tinyfrac"` added to lpjparams.js
- option `-area filename` added to the `printglobal` utility to scale with the cell areas read from the file specified.

### Changed

- Lake fraction and land-use fractions are rescaled by the reciprocal of land fraction. If lake fraction is greater than one, lake fraction is set to one and a warning is printed.
- Lakes can be enabled without river routing by setting `"with_lakes" : true`. Variable `dmass_lake` is always written to restart file.
- lake fraction is read in utility `lpjprint` and printed
- flag `"shuffle_spinup_climate"` only read if `"nspinup"` is non-zero.
- Terrestrial area and harvest fraction are read in utility `printharvest`. Output format changed to CSV.

## [5.6.27] - 2023-08-08

### Removed

- removed large chunks of code in `allocation_daily_crop.c` that was inactive

### Fixed

- fixed output `PFT_VEGC` to not include `bm_inc.nitrogen` for crops in `daily_agriculture.c`
- fixed global flux summation for `flux.soil.carbon` and `flux.soil.nitrogen` (`+=` instead of `=`), bug was introduced in version 5.6.25

## [5.6.26] - 2023-08-07

### changed

- avoid daily leaf turnover for biomass grass and harvest leaf turnover as dead biomass at harvest event instead (reduces litter and increases harvest)
- introduce harvest dates depending on green/brown harvest instead of C threshold for biomass grass
- literature-based C/N ratios and fn_turnover for biomass grass depending on green/brown harvest configuration
- results and literature references can be found in issue [#275](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/275).

### Added

- introduced `#define` flag to change from green harvest (default) to brown harvest: `#define BMGR_BROWN`
- two options include green harvest (fresh biomass, high N content - usually used for biofuels) and brown harvest (dry biomass, high lignin content, low N content - usually used for combustion)
- brown harvest uses later harvest dates (defined in lpjparam.js) and higher N recovery (as lower fn_turnover defined in pft.js) than green harvest
- the `#define` flag is required to read the correct flag-specific parameter from the lpjparam.js and pft.js, see issue [#311](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/311)

## [5.6.25] - 2023-07-29

### Added

- New outputs added for N fluxes from timber extraction and for N pools in wood product pools
- New outputs added for `estab_storage` C and N
- Added missing N fluxes, `estab_storage`, and reservoirs to `globalflux` output
- Added new fluxes to man pages

### Fixed 

- Corrected/added units in comments in output.h
- Corrected units in `flux_sum` man page
- Corrected `N2O_DENIT` output, which now no longer also includes `N2O_NIT` values
- Corrected writing of mineral N pools to outputs, so that also mineral N in frozen soil layers is included
- Corrected writing of `VEGN` for crops, so that `bm_inc.nitrogen` is no longer included to avoid double accounting

### Removed

- Removed unecessary initialization of output variables to zero in `update_annual.c`, which is done in `initoutputdata.c`


## [5.6.24] - 2023-07-29

### Fixed

-  The number of bands is now correctly checked for the `read_residue_data` input.


## [5.6.23] - 2023-07-28

### Changed

- Instead of the soil code the index of the grid cell is written into the NetCDF file for `"grid"` output.


## [5.6.22] - 2023-07-27

### Changed

- Configuration files `lpjml_netcdf.js`, `lpjml_non.js`, `param_non.js`, and `lpjparam_non.js` updated to latest lpjml version.

### Removed

- Predefined daily outputs (`"d_lai"`-`"d_pet"`) and corresponding functions removed.

- Version numbers removed from man pages.

### Fixed

- Variable `issocket` always initialized to `FALSE` in `readfilename.c`.


## [5.6.21] - 2023-07-21

### Changed

- Flag `"nitrogen_coupled"` disabled for lpjml offline runs. Water stress is always coupled with nitrogen stress. Only if lpjml is compiled with `-DCOUPLING_WITH_FMS` flag is enabled.

- Flag `"new_trf"` renamed to `"transp_suction_fcn"`.


## [5.6.20] - 2023-07-13

### Added

- New flags `"fix_co2"` and `"fix_co2_year"` added.

### Changed

- Flag `"istimber"` renamed to `"luc_timber"`.

- Flag `"new_phenology"` renamed to `"gsi_phenology"`.


src/lpj/readconfig.c
/**************************************************************************************/
/**                                                                                \n**/
/**                   r  e  a  d  c  o  n  f  i  g  .  c                           \n**/
/**                                                                                \n**/
/**     C implementation of LPJmL                                                  \n**/
/**                                                                                \n**/
/**     Function reads LPJ configuration file                                      \n**/
/**     Input is prepocessed by cpp                                                \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

Bool readconfig(Config *config,        /**< LPJ configuration */
                Pfttype scanfcn[],     /**< array of PFT-specific scan
                                             functions */
                int ntypes,            /**< Number of PFT classes */
                int nout,              /**< Maximum number of output files */
                int *argc,             /**< pointer to the number of arguments */
                char ***argv,          /**< pointer to the argument vector */
                const char *usage      /**< usage information or NULL */
               )                       /** \return TRUE on error */
{
  FILE *file;
  LPJfile *lpjfile;
  const char *s;
  Verbosity verbosity;
  const char *sim_id[]={"lpj","lpjml","lpjml_image","lpjml_fms"};
  config->arglist=catstrvec(*argv,*argc); /* store command line in arglist */
  config->coupled_model=NULL;
  file=openconfig(config,argc,argv,usage);
  if(file==NULL)
    return TRUE;
  verbosity=(isroot(*config)) ? config->scan_verbose : NO_ERR;
  lpjfile=parse_json(file,verbosity);
  if(config->nopp)
    fclose(file);
  else
    pclose(file);
  if(lpjfile==NULL)
    return TRUE;
  s=fscanstring(lpjfile,NULL,"sim_name",verbosity);
  if(s==NULL)
  {
    if(verbosity)
      fputs("ERROR121: Cannot read simulation name.\n",stderr);
    closeconfig(lpjfile);
    return TRUE;
  }
  config->sim_name=strdup(s);
  if(config->sim_name==NULL)
  {
    printallocerr("sim_name");
    closeconfig(lpjfile);
    return TRUE;
  }
  if(iskeydefined(lpjfile,"version"))
  {
    s=fscanstring(lpjfile,NULL,"version",verbosity);
    if(s==NULL)
    {
      if(verbosity)
        fputs("ERROR121: Cannot read version.\n",stderr);
      closeconfig(lpjfile);
      return TRUE;
    }
    if(verbosity && strncmp(s,LPJ_VERSION,strlen(s)))
      fprintf(stderr,"WARNING025: LPJ version '%s' does not match '" LPJ_VERSION "'.\n",s);
  }
  /* Read LPJ configuration */
  config->sim_id=LPJML;
  if(fscankeywords(lpjfile,&config->sim_id,"sim_id",sim_id,4,TRUE,verbosity))
  {
    closeconfig(lpjfile);
    return TRUE;
  }
#if defined IMAGE && defined COUPLED
  if(config->sim_id!=LPJML && config->sim_id!=LPJ && config->sim_id!=LPJML_IMAGE)
  {
    if(verbosity)
      fprintf(stderr,"ERROR123: Invalid simulation type, must be \"lpjml\" or \"lpj\" or \"lpjml_image\".\n");
    closeconfig(lpjfile);
    return TRUE;
  }
#else
  if(config->sim_id==LPJML_IMAGE)
  {
    if(verbosity)
      fputs("ERROR219: LPJmL has to be compiled with '-DIMAGE -DCOUPLED' for simulation type \"lpjml_image\".\n",stderr);
    closeconfig(lpjfile);
    return TRUE;
  }
#endif
  if(fscanconfig(config,lpjfile,scanfcn,ntypes,nout))
  {
    closeconfig(lpjfile);
    return TRUE;
  }
  closeconfig(lpjfile);
  return FALSE;
} /* of 'readconfig' */
